src/snowflake/snowpark/_internal/ast.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+import base64
+import itertools
+import sys
+import uuid
+from dataclasses import dataclass
+from typing import Callable, Tuple
+
+import snowflake.snowpark._internal.proto.generated.ast_pb2 as proto
+from snowflake.snowpark.version import VERSION
+
+# The current AST version number (generated by the DSL).
+CLIENT_AST_VERSION = proto.__Version__.MAX_VERSION
+
+
+@dataclass
```

Comment:
why this is a dataclass?

Reply:
I assume the question implies that a named tuple would work equally well. I have a very minor preference for data classes over named tuples, but not for internal types, it's a lot less important here.

I left things as they are for now. (But added a named tuple for the return value of `AstBatch.flush()`, so now we do things slightly inconsistently).

Reply:
namedtuple is a bad idea, unless you need the `is-a` tuple relationship - no need to use it.

Inheritance should be used carefully, as it comes with certain implications.

Can we use dataclass here to stay consistent?

Reply:
Oh my original question is why this needs to be immutable? I guess that's the reason to use `dataclass` right?

If it just a simple group of immutable items, I would prefer using namedtuple. If you are going to define methods then go with `dataclass`. Just a suggestion.

Reply:
It think it boils down to modeling here, each object has different contracts:

What is a tuple?
A tuple is fixed-length sequence of items (so, order matters!).
-> A namedtuple expands a tuple. So a namedtuple "is-a" tuple.

What is a dataclass?
A dataclass is a trait to mix into a class.

For context, cf.

https://peps.python.org/pep-0557/#rationale

> No base classes or metaclasses are used by Data Classes. Users of these classes are free to use inheritance and metaclasses without any interference from Data Classes. The decorated classes are truly “normal” Python classes. The Data Class decorator should not interfere with any usage of the class.

https://peps.python.org/pep-0557/#why-not-just-use-namedtuple

A namedtuple has a good chunk of disadvantages, my biggest pet peeve here is that it messes with the inheritance. Imho, inheritance should be used when it makes sense to specify relationships. Traits/mixins/interfaces are preferred when possible.

Reply:
Yes, if we don't need inheritance or just used it as a immutable message, namedtuple is fine.

Reply:
@sfc-gh-azhan the other way round :) Because we do not need inheritance and a relationship to tuple, imho we should use dataclass. Tbh, I do not think that this a very important detail. Yet, it would make sense to keep things succinct with @dataclass.

Reply:
I try to understand why the relationship with tuple is a problem?